filecoin-proofs/src/api/post_adapter.rs

```diff
@@ -0,0 +1,626 @@
+        ));
+    }
+
+    if proofs_len != replicas_c {
+        return Err(format_err!(
+            "LEN(proofs) must == CEIL(LEN(comm_rs)/2): {:?}, {:?}",
+            proofs_len,
+            replicas_c
+        ));
+    }
+
+    if faults_max
+        .map(|m| *m > commrs_len as u64 - 1)
+        .unwrap_or(false)
+    {
+        return Err(format_err!(
+            "MAX(faults) must <= LEN(comm_rs)-1: {:?}, {:?}",
+            faults_max,
+            commrs_len - 1
+        ));
+    }
+
+    let mut fixed: Vec<VerifyPoStFixedSectorsCountInput> = vec![];
+
+    let chunks = dynamic.comm_rs.chunks_exact(POST_SECTORS_COUNT);
+
+    let remainder = chunks.remainder();
+
+    // each POST_SECTORS_COUNT grouping of commitments maps to a single fixed
+    // sectors-count PoSt verification call
+    for (i, chunk) in chunks.enumerate() {
+        let mut comm_rs: [Commitment; POST_SECTORS_COUNT] = [[0; 32]; POST_SECTORS_COUNT];
+        for (i, comm_r) in chunk.iter().enumerate() {
+            comm_rs[i] = *comm_r;
+        }
+
+        fixed.push(VerifyPoStFixedSectorsCountInput {
+            sector_bytes: dynamic.sector_bytes,
+            comm_rs,
+            challenge_seed: dynamic.challenge_seed,
+            proof: dynamic.proofs[i],
+            faults: Vec::new(),
+        })
+    }
+
+    // If we receive a number of comm_rs which does not divide evenly by
+    // POST_SECTORS_COUNT, form a grouping of comm_rs which does divide
+    // evenly by POST_SECTORS_COUNT by duplicating the last comm_r in the
+    // group until LEN(comm_rs) == POST_SECTORS_COUNT.
+    if !remainder.is_empty() {
+        let mut comm_rs: [Commitment; POST_SECTORS_COUNT] = Default::default();
+
+        let vs = repeat(remainder[remainder.len() - 1]);
+        let rs = remainder.iter().cloned();
+        let it = rs.chain(vs).take(POST_SECTORS_COUNT).enumerate();
+
+        for (i, comm_r) in it {
+            comm_rs[i] = comm_r;
+        }
+
+        fixed.push(VerifyPoStFixedSectorsCountInput {
+            sector_bytes: dynamic.sector_bytes,
+            comm_rs,
+            challenge_seed: dynamic.challenge_seed,
+            proof: dynamic.proofs[proofs_len - 1],
+            faults: Vec::new(),
+        })
+    }
+
+    // Map the dynamic sectors count indices to the fixed sectors count indices.
+    for fault in dynamic.faults {
+        let i = (fault as f64 / POST_SECTORS_COUNT as f64).floor();
+        fixed[i as usize]
+            .faults
+            .push(fault % POST_SECTORS_COUNT as u64);
+    }
+
+    // If we had to duplicate commitments to make LEN(COMMITMENTS) ==
+    // POST_SECTORS_COUNT and the thing we duplicated was faulty, ensure that
+    // the duplicates are marked as faulty, too.
+    if !remainder.is_empty() {
+        let fixed_len = { fixed.len() - 1 };
+        let remdr_len = { remainder.len() };
+
+        let last_remainder_fault: Option<u64> =
+            { fixed[fixed_len].faults.get(remdr_len - 1).cloned() };
```

Comment:
Will the next to last `faults` always have `remdr_len - 1` items here? (just asking since I didn't quite followed the `faults` generation logic)

Reply:
I added an example to the code (in comments), which I'll repaste here:

```
// POST_SECTORS_COUNT = 2
//
// dynamic = {
//   comm_rs=[a, b, c]
//   faults=[2]
// }
//
// fixed = [
//   {comm_rs=[a, b] faults=[]}
//   {comm_rs=[c, c] faults=[0, 1]}
// ]
```

> Will the next to last faults always have remdr_len - 1 items here?

I don't quite understand your question. Would you mind rephrasing for me, please?